src/Native/Unix/System.Native/pal_io.c

```diff
@@ -45,6 +46,8 @@
 #define lstat_ lstat
 #endif
 
+#ifdef static_assert
```

Comment:
With newer Clang and GCC, `static_assert` is defined so the only case where we wouldn't compile these static asserts would be if you compile with an older version of Clang or GCC that do not support those.

Reply:
So should be define the static_assert in the common header if it is not defined? It would be better than ifdefing each use.

Reply:
https://github.com/dotnet/corefx/pull/25032/commits/15e76fb80b7995a42f66a8c2ca81dc53fa4fc3da defines `c_static_assert` which fixes your comment. It forces us to define `-Wno-typedef-redefinition`, which might be quite detrimental. IMO since `static_assert` is not defined only on older compilers, just gating for its presence brings the best tradeoff.